code/obj/machinery/sleeper.dm

```diff
@@ -128,146 +124,96 @@
+					src.time = clamp(src.time + (t*10), 0, 1800)
+				. = TRUE
+			if("inject")
+				if (src.our_sleeper && src.our_sleeper.occupant && !src.timing)
+					src.our_sleeper.inject(usr, TRUE)
+				. = TRUE
+			if("eject")
+				if (src.our_sleeper && src.our_sleeper.occupant)
+					src.our_sleeper.go_out()
+				. = TRUE
+
+	ui_data(mob/user)
+		if (!src.our_sleeper)
+			return list(
+				"sleeperGone" = TRUE,
+				"hasOccupant" = FALSE,
+			)
+
+		. = list(
+			"sleeperGone" = FALSE,
+			"hasOccupant" = FALSE,
+			"rejuvinators" = list(),
+			"recharging" = src.our_sleeper.no_med_spam && world.time < src.our_sleeper.no_med_spam + 50,
+			"isTiming" = src.timing,
+			"time" = src.time,
+			"timeStarted" = src.time_started,
+			"timeNow" = TIME,
+		)
 
-		// Capped at 3 min. Used to be 10 min, Christ.
-		if (href_list["tp"])
-			if (src.our_sleeper && src.time < 1800)
-				var/t = text2num(href_list["tp"])
-				if (t > 0 && src.timing && src.our_sleeper.occupant)
-					// People do use sleepers for grief from time to time.
-					logTheThing("station", usr, src.our_sleeper.occupant, "increases a sleeper's timer ([src.our_sleeper.emagged ? "<b>EMAGGED</b>, " : ""]occupied by [constructTarget(src.our_sleeper.occupant,"station")]) by [t] seconds at [log_loc(src.our_sleeper)].")
-				//src.time = min(180, max(0, src.time + t))
-				src.time = clamp(src.time + (t*10), 0, 1800)
-
-		if (href_list["rejuv"])
-			if (src.our_sleeper && src.our_sleeper.occupant)
-				if (src.timing)
-					// So they can't combine this with manual injections to spam/farm reagents.
-					usr.show_text("Occupant alarm clock active. Manual injection unavailable.", "red")
-				else
-					src.our_sleeper.inject(usr, 1)
-
-		if (href_list["refresh"])
-			if (!src.our_sleeper && src.find_sleeper_in_range)
-				our_sleeper = locate() in orange(src,1)
-			else if (istype(src.loc, /obj/machinery/sleeper))
-				our_sleeper = src.loc
-
-		if (href_list["eject_occupant"])
-			if (src.our_sleeper && src.our_sleeper.occupant)
-				src.our_sleeper.go_out()
-				src.remove_dialog(usr)
-				usr.Browse(null, "window=sleeper")
-
-		if (istype(src, /obj/machinery/sleep_console/portable))
-			src.attack_hand(usr)
-		else
-			src.updateUsrDialog()
-		return
+		var/mob/occupant = src.our_sleeper.occupant
+		if (occupant)
+			. += list(
+				"hasOccupant" = TRUE,
+				"occupantStat" = occupant.stat,
+				"health" = occupant.health,
+				"oxyDamage" = occupant.get_oxygen_deprivation(),
+				"toxDamage" = occupant.get_toxin_damage(),
+				"burnDamage" = occupant.get_burn_damage(),
+				"bruteDamage" = occupant.get_brute_damage(),
+			)
+
+			// We don't have a fully-fledged reagent scanner built-in. Of course, this also means
+			// we can't detect our own poisons if the sleeper's emagged. Too bad.
+			for (var/R in occupant.reagents.reagent_list)
+				var/datum/reagent/medical/MR = occupant.reagents.reagent_list[R]
+				if (istype(MR))
```

Comment:
Sanity check from someone better at DM than me, would this be better written as below?
```
			for (var/datum/reagent/medical/MR in occupant.reagents.reagent_list)
```

Reply:
No that's different behaviour. `R` is a key and `MR` is a value in the associative list in original code. Your version would still iterate over keys (which are reagent *ids* and not reagents).

Reply:
Oh, duh. This is why I ask for nerds.